### PR TITLE
User Interface Rework

### PR DESCRIPTION
**Pull request by AMastryukov**
- Overhaul the entire user interface implementation

### CODE REVIEW

File: code/items/weapons/TTTWeapon.cs [hunk 1]
```diff
@@ -7,17 +7,6 @@
...
-    // public class WeaponAttribute : LibraryAttribute
-    // {
-    //     public WeaponType WeaponType;
-
-    //     public WeaponAttribute(string name) : base(name)
-    //     {
-
-    //     }
-    // }
-
```
Comment (Alf21) [inline code/items/weapons/TTTWeapon.cs:20]: Please keep it as this should be added by sbox soonTM (so we gonna be able to fetch data without the need initializing and spawning such a weapon),

File: code/player/TTTPlayer.Confirmation.cs [hunk 2]
```diff
@@ -47,51 +47,44 @@ private void TickAttemptInspectPlayerCorpse()
...
-
-                        // TODO Handling if a player disconnects!
-                        if (playerCorpse.Player != null && playerCorpse.Player.IsValid())
-                        {
-                            playerCorpse.Player.IsConfirmed = true;
-                            playerCorpse.Player.CorpseConfirmer = this;
-
-                            int credits = playerCorpse.Player.Credits;
+                    InspectMenu.Instance.Enabled = false;
+                    return;
```
Comment (Alf21) [inline code/player/TTTPlayer.Confirmation.cs:53]:
```suggestion

                    return;
```

File: code/ui/Crosshair.cs [hunk 3]
```diff
@@ -3,166 +3,154 @@
...
-        private float ChargeTime;
-
-        public Crosshair()
+        private bool _showDot = true;
+        public bool ShowDot
         {
-            Current = this;
-            StyleSheet.Load("/ui/Crosshair.scss");
+            get { return _showDot; }
+            set { _showDot = value; UpdateCrosshair(); }
```
Comment (Alf21) [inline code/ui/Crosshair.cs:16]: format

File: code/ui/Crosshair.cs [hunk 4]
```diff
@@ -3,166 +3,154 @@
...
-            }
-        }
-
         public static Crosshair Current;
 
-        private Panel ChargeBar;
-        private float ChargeTime;
-
-        public Crosshair()
+        private bool _showDot = true;
```
Comment (Alf21) [inline code/ui/Crosshair.cs:12]: Please put this below the public property for code consistency

Comment (Alf21) [reply code/ui/Crosshair.cs:12]: Same below.

File: code/ui/Hud.cs [hunk 5]
```diff
@@ -33,20 +35,54 @@ public static void OnHotReloaded()
...
+        }
+
+        public class GeneralHud : Panel
+        {
+            public GeneralHud(Sandbox.UI.Panel parent) : base(parent)
             {
                 Parent = parent;
 
-                Parent.AddChild<PlayerInfo>();
+                Parent.AddChild<RadarDisplay>();
```
Comment (Alf21) [inline code/ui/Hud.cs:74]: I'd have kept it separated and in the radar item file to support code-separation.

File: code/ui/alivehud/quickshop/QuickShopItem.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        public void SetItem(ShopItemData buyableItemData)
+        {
+            ItemData = buyableItemData;
+
+            _itemNameLabel.Text = $"{buyableItemData.Name}";
+            _itemPriceLabel.Text = $"${buyableItemData.Price}";
+
+            var icon = Texture.Load($"/ui/weapons/{buyableItemData.Name}.png");
```
Comment (Alf21) [inline code/ui/alivehud/quickshop/QuickShopItem.cs:49]: Don't use `var`

File: code/ui/components/generic/Panel.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        public bool Enabled
+        {
+            get => _isEnabled;
+            set
+            {
+                _isEnabled = value;
+
+                SetClass("disabled", !_isEnabled);
+            }
+        }
```
Comment (Alf21) [inline code/ui/components/generic/Panel.cs:15]:
```suggestion
        public bool Enabled
        {
            get => _isEnabled;
            set
            {
                _isEnabled = value;

                SetClass("disabled", !_isEnabled);
            }
        }
        protected bool _isEnabled = true;
```

File: code/ui/components/generic/Window.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            Header = new(this);
+            Header.AddClass("header");
+            Header.AddClass("rounded-top");
+            Header.AddClass("background-color-secondary");
+
+            WindowLabel = Header.Add.Label("Window");
+
+            CloseButton = Header.Add.Button("✕");
+            CloseButton.AddClass("button");
+            CloseButton.AddEventListener("onclick", () => { Enabled = false; });
```
Comment (Alf21) [inline code/ui/components/generic/Window.cs:29]: Multiline please :)

File: code/ui/generalhud/chat/ChatBox.cs [hunk 9]
```diff
@@ -49,136 +68,179 @@ public override void Tick()
...
-            _input.Text = "";
+        public void AddEntry(string header, string content, Channel channel, string avatar = null, string teamName = null)
+        {
+            _lastChatFocus = 0f;
 
-            if (string.IsNullOrWhiteSpace(msg))
+            if (channel == Channel.Team && string.IsNullOrEmpty(teamName))
             {
+                Log.Error("Cannot add chat entry to Team channel without a team name.");
                 return;
```
Comment (Alf21) [inline code/ui/generalhud/chat/ChatBox.cs:162]:
```suggestion

                return;
```

File: code/ui/generalhud/chat/ChatBox.cs [hunk 10]
```diff
@@ -49,136 +68,179 @@ public override void Tick()
...
 
-            if (wasTeamChatting)
-            {
-                SayTeam(msg);
-            }
-            else
+            if (header == null)
             {
-                Say(msg);
+                header = "";
```
Comment (Alf21) [inline code/ui/generalhud/chat/ChatBox.cs:167]: CAn be a oneliner: `header ??= "";` (null-coalescing-operator)

File: code/ui/generalhud/inspectmenu/InspectEntry.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            _inspectQuickLabel = Add.Label();
+            _inspectQuickLabel.AddClass("quick-label");
+        }
+
+        public void SetData(string imagePath, string description)
+        {
+            ImagePath = imagePath;
+            Description = description;
+
+            var icon = Texture.Load(imagePath);
```
Comment (Alf21) [inline code/ui/generalhud/inspectmenu/InspectEntry.cs:35]: `var`

File: code/ui/generalhud/inspectmenu/InspectEntry.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+            if (icon == null)
+            {
+                icon = Texture.Load($"/ui/none.png");
```
Comment (Alf21) [inline code/ui/generalhud/inspectmenu/InspectEntry.cs:38]: One liner (null-coalescing-operator). Could be included into the panelbackground declaration below as well.

File: code/ui/generalhud/inspectmenu/InspectMenu.cs [hunk 13]
```diff
@@ -9,278 +9,215 @@
...
+            _perkEntries.Clear();
 
-                public InspectItem(Panel parent)
+            // Populate perk entries
+            if (perks != null)
+            {
+                foreach (string perkName in perks)
                 {
-                    Parent = parent;
+                    var perkEntry = new InspectEntry(this);
```
Comment (Alf21) [inline code/ui/generalhud/inspectmenu/InspectMenu.cs:187]: `var`

File: code/ui/generalhud/inspectmenu/InspectMenu.cs [hunk 14]
```diff
@@ -9,278 +9,215 @@
...
-                    }
+        public override void Tick()
+        {
+            if (Enabled && _playerCorpse?.Transform.Position.Distance(Local.Pawn.Position) > 100f)
+            {
+                Enabled = false;
+            }
 
-                    _footerLabel.SetClass("hide", false);
+            var timeSinceDeath = Globals.Utils.TimerString(Time.Now - _confirmationData.Time);
```
Comment (Alf21) [inline code/ui/generalhud/inspectmenu/InspectMenu.cs:214]: `var`

File: code/ui/generalhud/inventorywrapper/effects/Effect.cs [hunk 15]
```diff
@@ -38,39 +32,62 @@ private set
                 }
                 else
                 {
-                    Countdown?.Delete();
-                    Countdown ??= null;
+                    _countdownLabel?.Delete();
+                    _countdownLabel ??= null;
```
Comment (Alf21) [inline code/ui/generalhud/inventorywrapper/effects/Effect.cs:36]: Just htought about it and it really doesn't make much sense, as it does `_countdownLabel = _countdownLabel ?? null`

File: code/ui/generalhud/inventorywrapper/effects/Effects.scss [hunk 16]
```diff
@@ -1,47 +1,44 @@
...
 
-    Effect {
-        border: 1px solid #031421;
-        background-color: rgba(0, 0, 0, 0.7);
-        border-radius: 4px;
-        width: 62px;
-        height: 62px;
-        margin-left: 10px;
-        margin-top: 10px;
+Effect {
```
Comment (Alf21) [inline code/ui/generalhud/inventorywrapper/effects/Effects.scss:8]: Why isn't the effect in the effects class anymore?

File: code/ui/generalhud/nameplate/Nameplate.cs [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            foreach (HealthGroup healthGroup in HealthGroupList)
+            {
+                if (health >= healthGroup.MinHealth)
+                {
+                    return healthGroup;
+                }
+            }
+
+            return HealthGroupList[HealthGroupList.Length - 1];
```
Comment (Alf21) [inline code/ui/generalhud/nameplate/Nameplate.cs:69]:
```suggestion
            return HealthGroupList[^1];
```

File: code/ui/generalhud/playerinfo/PlayerInfoDisplay.cs [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+        public override void Tick()
+        {
+            base.Tick();
+
+            if (Local.Pawn is not TTTPlayer player)
+            {
+                return;
+            }
+
+            Enabled = Local.Pawn is TTTPlayer && !player.IsSpectator && !player.IsSpectatingPlayer;
```
Comment (Alf21) [inline code/ui/generalhud/playerinfo/PlayerInfoDisplay.cs:48]: So you removed first person spectator support?

File: code/ui/generalhud/voicechat/VoiceChatDisplay.cs [hunk 19]
```diff
@@ -0,0 +1,26 @@
+using System.Linq;
+
+using Sandbox;
+using Sandbox.UI;
+
+namespace TTTReborn.UI
+{
+    public class VoiceChatDisplay : Panel
+    {
+        public static VoiceChatDisplay Current { get; internal set; }
```
Comment (Alf21) [inline code/ui/generalhud/voicechat/VoiceChatDisplay.cs:10]: `Instance`

File: code/ui/generalhud/voicechat/VoiceChatEntry.cs [hunk 20]
```diff
@@ -68,26 +80,6 @@ public override void Tick()
...
-                if (player.LifeState == LifeState.Dead)
-                {
-                    Style.BackgroundColor = _deadColor.WithAlpha(0.9f);
-                }
-                else if (player.IsTeamVoiceChatEnabled && player.Role is not NoneRole)
-                {
-                    Style.BackgroundColor = player.Role.Color.WithAlpha(0.9f);
-                }
-            }
-
```
Comment (Alf21) [inline code/ui/generalhud/voicechat/VoiceChatEntry.cs:90]: Did you removed the player state indicators?

**Review by Alf21 — COMMENTED** [review]
I really like what you did, so great changes. Most of the comments are just code-style things for consistency. In some rare cases, it seems like you removed already implemented core functions, that probably should be replaced before merging?

Additionally, you added the file `materials/ammo/9mmbox.vmat_c `, intended?

File: code/global/RPCs.cs [hunk 21]
```diff
@@ -117,6 +113,15 @@ public static void ClientConfirmPlayer(TTTPlayer confirmPlayer, PlayerCorpse pla
...
 
+        [ClientRpc]
+        public static void ClientEnableInspectMenu(PlayerCorpse playerCorpse)
+        {
+            if (!InspectMenu.Instance.Enabled)
+            {
+                InspectMenu.Instance.InspectCorpse(playerCorpse);
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (AMastryukov) [reply code/global/RPCs.cs:124]: I don't think we do, this was a mistake on my end. I actually had someone join and make my menu pop up when they inspected the body. This was probably the reason. Good catch!

File: code/items/perks/radar/RadarPoint.cs [hunk 22]
```diff
@@ -32,17 +37,28 @@ public override void Tick()
...
                 Style.Left = Length.Fraction(screenPos.x);
                 Style.Top = Length.Fraction(screenPos.y);
                 Style.Dirty();
             }
         }
     }
+
+    public class RadarDisplay : Panel
+    {
+        public static RadarDisplay Current { get; set; }
```
[the remark this answers is not in the document]
Comment (AMastryukov) [reply code/items/perks/radar/RadarPoint.cs:56]: Yes, I forgot to change it.

File: code/items/weapons/TTTWeapon.cs (hunk 1 above)
Comment (AMastryukov) [reply code/items/weapons/TTTWeapon.cs:20]: I'm a simple man - I see comment, I delete it 🤣 
I'll revert it and add additional clarification

File: code/ui/generalhud/nameplate/Nameplate.cs (hunk 17 above)
Comment (AMastryukov) [reply code/ui/generalhud/nameplate/Nameplate.cs:69]: > Additionally, you added the file `materials/ammo/9mmbox.vmat_c `, intended?

The vmats are the worst. They appear sometimes and when making a lot of changes they just slip by and end up in the commit. I'm just glad only one made it through given that I had so many commits made.

File: code/ui/generalhud/inventorywrapper/effects/Effects.scss (hunk 16 above)
Comment (AMastryukov) [reply code/ui/generalhud/inventorywrapper/effects/Effects.scss:8]: I made everything separate classes. Too many things had nested classes and they were annoying to work with, resulting in stuff like `Parent.Parent.Parent`. I figured everything should be in the same scope and tried to maintain consistency even if I didn't absolutely need to.

File: code/ui/generalhud/playerinfo/PlayerInfoDisplay.cs (hunk 18 above)
Comment (AMastryukov) [reply code/ui/generalhud/playerinfo/PlayerInfoDisplay.cs:48]: This was a mistake.

File: code/ui/generalhud/voicechat/VoiceChatEntry.cs (hunk 20 above)
Comment (AMastryukov) [reply code/ui/generalhud/voicechat/VoiceChatEntry.cs:90]: I had reworked this part when fixing a bug and forgot to add the spectator colors.

File: code/ui/generalhud/playerinfo/PlayerRoleDisplay.cs [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+        public override void Tick()
+        {
+            base.Tick();
+
+            if (Local.Pawn is not TTTPlayer player)
+            {
+                return;
+            }
+
+            Enabled = Local.Pawn is TTTPlayer && !player.IsSpectator && !player.IsSpectatingPlayer && Gamemode.Game.Instance.Round is Rounds.InProgressRound;
```
Comment (Alf21) [inline code/ui/generalhud/playerinfo/PlayerRoleDisplay.cs:36]: Before, the RoleName was replaced with the currently spectated player name when spectating another player in firstperson (or thirdperson)

File: code/items/perks/radar/Radar.cs [hunk 24]
```diff
@@ -55,7 +55,7 @@ private void UpdatePositions()
 
                 foreach (Vector3 vector3 in _lastPositions)
                 {
-                    _cachedPoints.Add(new RadarPoint(vector3));
+                    _cachedPoints.Add(new RadarPoint(vector3 + Vector3.Up * 45));
```
Comment (mzegar) [inline code/items/perks/radar/Radar.cs:58]: Could maybe make this a const static `RADARPOINT_UI_OFFSET`

File: code/items/perks/radar/RadarPoint.cs [hunk 25]
```diff
@@ -18,9 +18,14 @@ public RadarPoint(Vector3 vector3)
 
             StyleSheet.Load("/items/perks/radar/RadarPoint.scss");
 
-            Hud.Current.RootPanel.AddChild(this);
+            // Hud.Current.RootPanel.AddChild(this);
```
Comment (mzegar) [inline code/items/perks/radar/RadarPoint.cs:21]: Can this go?

File: code/items/weapons/NewtonLauncher.cs [hunk 26]
```diff
@@ -109,18 +109,7 @@ public override void CreateHudElements()
...
-                false,
-                10,
-                2,
-                0,
-                0,
-                NEWTON_CHARGE_TIME,
-                Color.Green));
-            CrosshairPanel.Parent = Local.Hud;
-            CrosshairPanel.AddClass(ClassInfo.Name);
+            // TODO: Create a special HUD element for Newton Launcher
```
Comment (mzegar) [inline code/items/weapons/NewtonLauncher.cs:112]: Let's make an issue for this.

File: code/items/weapons/TTTWeapon.cs [hunk 27]
```diff
@@ -7,6 +7,9 @@
 
 namespace TTTReborn.Items
 {
+    // DO NOT DELETE
+    // This should be added by sbox soonTM (so we gonna be able to fetch data without the need initializing and spawning such a weapon)
+    //
```
Comment (mzegar) [inline code/items/weapons/TTTWeapon.cs:12]: Another issue for this imo

File: code/player/TTTPlayer.Hints.cs [hunk 28]
```diff
@@ -21,29 +21,37 @@ private void TickEntityHints()
...
             }
 
-            //If we are looking at a target and don't have a current hint, let's see if we can make one.
+            // If we are looking at a target and don't have a current hint, let's see if we can make one.
             if (target != null)
             {
                 if (target.CanHint(player) && _currentHint == null)
                 {
-                    _currentHint = target.DisplayHint(player); //Retrieves panel information from entity
+                    _currentHint = target.DisplayHint(player);
```
Comment (mzegar) [inline code/player/TTTPlayer.Hints.cs:40]: There was a bug previously I fixed 
https://github.com/TTTReborn/tttreborn/commit/0af6c6fd968f56cd081e0f03e3bad11a03dbec7f

It still may occur here

File: code/ui/Hud.cs (hunk 5 above)
Comment (AMastryukov) [reply code/ui/Hud.cs:74]: https://github.com/TTTReborn/tttreborn/issues/219

File: code/ui/generalhud/inventorywrapper/effects/Effect.cs (hunk 15 above)
Comment (AMastryukov) [reply code/ui/generalhud/inventorywrapper/effects/Effect.cs:36]: This part was probably edited by me when refactoring names, so I didn't bother to look.

File: code/ui/generalhud/playerinfo/PlayerRoleDisplay.cs (hunk 23 above)
Comment (AMastryukov) [reply code/ui/generalhud/playerinfo/PlayerRoleDisplay.cs:36]: Let's keep it hidden for now, and decide whether we want a mode where spectators don't know the roles of living players.

File: code/player/TTTPlayer.Hints.cs (hunk 28 above)
Comment (AMastryukov) [reply code/player/TTTPlayer.Hints.cs:40]: Good catch! I added your changes in.

**Review by mzegar — APPROVED** [review]
My changes were addressed.

**Review by Alf21 — APPROVED** [review]
Thanks!

**Pull request merged**